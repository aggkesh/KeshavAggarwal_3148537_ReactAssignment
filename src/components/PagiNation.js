import React from 'react';
import { Pagination } from 'react-bootstrap';
import { useHistory  } from 'react-router-dom';

function PagiNation({ productData, fetchProducts }) {

    const history = useHistory();
    const paginationItems = paginationItemsGenerator(productData, fetchProducts, history);
    const pages = numberOfPages(productData);

    return(
        <Pagination size="lg" className="d-flex justify-content-center m-4">
            <Pagination.Prev onClick={() => fetchProducts(productData.search, productData.activePage - 1, productData.sort, history) } disabled={ productData.activePage === 1 }/>
            { paginationItems }
            <Pagination.Next  onClick={() => fetchProducts(productData.search, productData.activePage + 1, productData.sort, history) } disabled={ productData.activePage === pages } />
        </Pagination>
    )
}

/**
 * Generate Diffent pages data
 * 
 * @param {*} productData populate products
 * @param {*} fetchProducts fuction to fetch products
 * @param {*} history history to move between app
 */
export const paginationItemsGenerator = (productData, fetchProducts, history) => {
    const paginationItems = [];
    const pages = numberOfPages(productData);

    for (let number = 1; number <= pages; number++) {
        paginationItems.push(
            <Pagination.Item key={ number } active={ number === productData.activePage } 
                onClick={() => fetchProducts(productData.search, number, productData.sort, history) } >
                { number }
            </Pagination.Item>,
        );
    }

    return paginationItems;
}

/**
 * Number of pages for pagination
 * 
 * @param {*} productData products to determine number of pages
 */
export const numberOfPages = (productData) => {
    return  Math.ceil(Number(productData.productTotalCount) / 8);
}

export default PagiNation