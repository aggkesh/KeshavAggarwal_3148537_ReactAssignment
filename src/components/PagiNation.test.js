import React from 'react';
import PagiNation, { numberOfPages, paginationItemsGenerator } from './PagiNation';

import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() })

describe('Pagination Snapshot Test Cases', () => {
    it('renders correctly enzyme', () => {
        const wrapper = shallow(<PagiNation productData={ [] } />)
      
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    
    it('number of pages calcualted correctly', () => {      
        const productData = {
            productTotalCount: 8
        }

        expect(numberOfPages(productData)).toEqual(1);

        productData.productTotalCount = 16
        expect(numberOfPages(productData)).toEqual(2);

        productData.productTotalCount = 24
        expect(numberOfPages(productData)).toEqual(3);

        productData.productTotalCount = 32
        expect(numberOfPages(productData)).toEqual(4);
    });

    it('number of pagination items to render correctly', () => {      
        const productData = {
            productTotalCount: 8
        }

        expect(paginationItemsGenerator(productData).length).toEqual(1);

        productData.productTotalCount = 16
        expect(paginationItemsGenerator(productData).length).toEqual(2);

        productData.productTotalCount = 24
        expect(paginationItemsGenerator(productData).length).toEqual(3);

        productData.productTotalCount = 32
        expect(paginationItemsGenerator(productData).length).toEqual(4);
    });
})