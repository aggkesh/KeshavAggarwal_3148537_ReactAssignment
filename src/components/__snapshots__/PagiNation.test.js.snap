// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Snapshot Test Cases renders correctly enzyme 1`] = `
<ForwardRef
  className="d-flex justify-content-center m-4"
  size="lg"
>
  <Prev
    disabled={false}
    onClick={[Function]}
  />
  <Next
    disabled={false}
    onClick={[Function]}
  />
</ForwardRef>
`;
